### PR TITLE
Make cache_filter caches not have to be statically allocated

### PR DESCRIPTION
**Pull request by ravenblackx**
Commit Message: Make cache_filter caches not have to be statically allocated
Additional Description: The previous implementation of the http cache filter returned caches only by reference, which meant they had to be owned by the (static) factory. Any config change that required changing the instantiation of the cache (or disabled the cache) would likely leak the existing cache. This change makes it so the cache is bound to the config, and may (or may not) be removed if the config changes, depending on if the new configuration will still use the same cache instance.
Risk Level: Small. cache filter is still tagged WIP.
Testing: Passes existing tests after change.
Docs Changes: n/a
Release Notes: http cache filter getCache interface changed from returning a reference to returning a shared_ptr - any third-party implementations of this interface will need to be updated accordingly. See [PR21114](https://github.com/envoyproxy/envoy/pull/21114) for example.
Platform Specific Features: n/a

### CODE REVIEW

File: docs/root/version_history/current.rst [hunk 1]
```diff
@@ -11,6 +11,10 @@ Minor Behavior Changes
 ----------------------
 *Changes that may cause incompatibilities for some users, but should not for most*
 
+* http cache filter: http cache filter `getCache` interface changed from returning a reference to
+  returning a shared_ptr - any third-party implementations of this interface will need to be
+  updated accordingly. See changes to `simple_http_cache.cc` and `simple_http_cache.h` in
+  [PR21114](https://github.com/envoyproxy/envoy/pull/21114) for example.
```
Comment (mattklein123) [inline docs/root/version_history/current.rst:17]: I don't think this markdown link syntax will render correctly, you need to use sphinx/RST link syntax. See other examples or online docs. Also check formatting CI in general.

**Comment by mattklein123** [comment]
Sorry needs a main merge to move over to the new YAML release note format (no more alpha order!)

/wait

**Comment by ravenblackx** [comment]
Again apologies for force-push. Things got out of sync as part of the merge in a way that made it reluctant to push or be fixed in any sane way. The only thing I touched manually was the yaml file.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**